zest/releaser/vcs.py

```diff
@@ -204,33 +246,41 @@ def _update_version(self, version):
             setup_version = self.get_setup_py_version()
             if not setup_version or (setup_version ==
                                      self.get_version_txt_version()):
-                open(versionfile, 'w').write(version + '\n')
-                logger.info("Changed %s to %r", versionfile, version)
+                open(versionfile, 'w').write(version + u'\n')
+                logger.info(
+                    u"Changed {0} to {1!r}".format(versionfile, version)
+                    )
                 return
 
-        good_version = "version = '%s'" % version
-        line_number = 0
-        setup_lines = open('setup.py').read().split('\n')
-        for line_number, line in enumerate(setup_lines):
-            match = VERSION_PATTERN.search(line)
-            if match:
-                logger.debug("Matching version line found: %r", line)
-                if line.startswith(' '):
-                    # oh, probably '    version = 1.0,' line.
-                    indentation = line.split('version')[0]
-                    # Note: no spaces around the '='.
-                    good_version = indentation + "version='%s'," % version
-                setup_lines[line_number] = good_version
-                contents = '\n'.join(setup_lines)
-                open('setup.py', 'w').write(contents)
-                logger.info("Set setup.py's version to %r", version)
-                return
+        content = read_text_file(filename)
+        tree = ast.parse(content, filename)
+        assignment = self._find_assignment(tree, u'version')
+        if assignment:
+            # This assumes that the version string does not span more than
+            # one line
+            position = (
+                assignment.value.lineno - 1,
+                assignment.value.col_offset,
+                assignment.value.col_offset + len(repr(assignment.value.s))
+                )
+            lines = content.splitlines()
+            logger.debug(u"Matching version line found: {0!r}".format(
+                lines[position[0]]))
+            lines[position[0]] = (
+                lines[position[0]][:position[1]] +
+                repr(version) +
+                lines[position[0]][position[2]:]
+                )
+            contents = u'\n'.join(lines)
+            open('setup.py', 'w').write(contents)
+            logger.info("Set setup.py's version to {0}".format(version))
+            return
```

Comment:
The comma should be at ``position[2]``. That said, this does rely on ``repr(value)`` being consistent and it may not be (``u'1.0'`` vs ``'1.0`` in Python 3; ``'1.0'`` vs ``b'1.0'`` in Python 2 - it could also use ``r'1.0'`` and other combinations, too which ``repr`` would render differently). I need to work in a regexp here as well to combat this, but using the AST to find the starting position is still a worthwhile change.

Thanks for prompting me to rethink this.